File: UnityMcpBridge/Editor/Tools/AddStep.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Reflection;
using System.Text.RegularExpressions;
using Newtonsoft.Json.Linq;
using UnityEditor;
using UnityEngine;

namespace MCPForUnity.Editor.Tools
{
    /// <summary>
    /// Generic step handler that can create different types of steps based on input parameters.
    /// Acts as a router to specific step creation handlers.
    /// </summary>
    public static class AddStep
    {
        // --- Step Type Registry ---
        private static readonly Dictionary<string, StepTypeHandler> StepHandlers = new Dictionary<string, StepTypeHandler>
        {
            { "delay", new StepTypeHandler("DelayStep", CreateDelayStep, new[] { "delay", "wait", "pause", "sleep" }) },
            { "click", new StepTypeHandler("ClickStep", CreateClickStep, new[] { "click", "tap", "press", "button" }) },
            // Add more step types here as needed
        };

        // --- Pattern Recognition ---
        private static readonly Regex GraphPattern = new Regex(@"(?:in|graph|asset:?\s*)?([\w-]+\.asset)", RegexOptions.IgnoreCase);
        private static readonly Regex DelayPattern = new Regex(@"(?:wait|delay|for:?\s*)?(\d+(?:\.\d+)?)\s*(?:seconds?|s)?", RegexOptions.IgnoreCase);
        private static readonly Regex ClickPattern = new Regex(@"(?:click|tap|press)\s+(?:on\s+)?([\w-]+)(?:/([^/\s]+))?", RegexOptions.IgnoreCase);
        private static readonly Regex StepTypePattern = new Regex(@"(?:add|create)\s+(?:a\s+)?(\w+)\s*(?:step)?", RegexOptions.IgnoreCase);

        // --- Cached Types ---
        private static Type nodeGraphType;
        private static Type stepsGraphType;
        private static Dictionary<string, Type> stepTypes = new Dictionary<string, Type>();

        // --- Main Handler ---
        public static object HandleCommand(JObject @params)
        {
            try
            {
                Debug.Log($"AddStep received parameters: {@params?.ToString()}");

                // Extract parameters
                string prompt = @params["prompt"]?.ToString();
                string stepType = @params["stepType"]?.ToString()?.ToLower();
                string targetGraphPath = @params["targetGraphPath"]?.ToString() ?? @params["path"]?.ToString();

                Debug.Log($"Initial parameters - stepType: {stepType}, targetGraphPath: {targetGraphPath}, prompt: {prompt}");

                // Determine step type from prompt if not explicitly provided
                if (string.IsNullOrEmpty(stepType) && !string.IsNullOrEmpty(prompt))
                {
                    stepType = DetermineStepTypeFromPrompt(prompt);
                    Debug.Log($"Determined stepType from prompt: {stepType}");
                }

                // Parse graph path from prompt if not provided
                if (string.IsNullOrEmpty(targetGraphPath) && !string.IsNullOrEmpty(prompt))
                {
                    var graphMatch = GraphPattern.Match(prompt);
                    if (graphMatch.Success)
                    {
                        targetGraphPath = $"Assets/_MCP/AddStepsExample/{graphMatch.Groups[1].Value}";
                        Debug.Log($"Parsed targetGraphPath from prompt: {targetGraphPath}");
                    }
                }

                // Validate required parameters
                if (string.IsNullOrEmpty(stepType))
                {
                    return CreateErrorResponse($"Could not determine step type. Please specify 'stepType' parameter or use keywords like: {string.Join(", ", GetAllStepKeywords())}");
                }

                if (string.IsNullOrEmpty(targetGraphPath))
                {
                    return CreateErrorResponse("targetGraphPath or path parameter is required.");
                }

                // Find the appropriate handler
                if (!StepHandlers.TryGetValue(stepType, out var handler))
                {
                    var availableTypes = string.Join(", ", StepHandlers.Keys);
                    return CreateErrorResponse($"Unknown step type: '{stepType}'. Available types: {availableTypes}");
                }

                // Normalize and validate the graph path
                targetGraphPath = SanitizeAssetPath(targetGraphPath);
                var loadedGraph = AssetDatabase.LoadAssetAtPath<ScriptableObject>(targetGraphPath);
                
                if (loadedGraph == null || !IsStepsGraph(loadedGraph))
                {
                    return CreateErrorResponse($"StepsGraph asset not found at path: {targetGraphPath}. Make sure the path is relative to Assets/ and uses forward slashes.");
                }

                // Call the specific handler
                var result = handler.CreateStep(loadedGraph, @params, prompt);
                
                if (result != null)
                {
                    return CreateSuccessResponse(
                        $"{handler.TypeName} added successfully to '{targetGraphPath}'.",
                        result
                    );
                }
                else
                {
                    return CreateErrorResponse($"Failed to create {handler.TypeName}");
                }
            }
            catch (Exception ex)
            {
                Debug.LogError($"Error in AddStep.HandleCommand: {ex.Message}\n{ex.StackTrace}");
                return CreateErrorResponse($"Exception occurred: {ex.Message}");
            }
        }

        // --- Response Helper Methods ---
        
        private static object CreateErrorResponse(string message)
        {
            return new Dictionary<string, object>
            {
                { "success", false },
                { "error", message }
            };
        }

        private static object CreateSuccessResponse(string message, object data = null)
        {
            var response = new Dictionary<string, object>
            {
                { "success", true },
                { "message", message }
            };
            
            if (data != null)
            {
                response["data"] = data;
            }
            
            return response;
        }

        // --- Step Type Determination ---
        private static string DetermineStepTypeFromPrompt(string prompt)
        {
            if (string.IsNullOrEmpty(prompt))
                return null;

            string lowerPrompt = prompt.ToLower();

            // Check each handler's keywords
            foreach (var kvp in StepHandlers)
            {
                if (kvp.Value.Keywords.Any(keyword => lowerPrompt.Contains(keyword)))
                {
                    return kvp.Key;
                }
            }

            // Fallback: try to extract from "add [type] step" pattern
            var stepTypeMatch = StepTypePattern.Match(prompt);
            if (stepTypeMatch.Success)
            {
                string extractedType = stepTypeMatch.Groups[1].Value.ToLower();
                if (StepHandlers.ContainsKey(extractedType))
                {
                    return extractedType;
                }
            }

            return null;
        }

        private static string[] GetAllStepKeywords()
        {
            return StepHandlers.SelectMany(kvp => kvp.Value.Keywords).Distinct().ToArray();
        }

        // --- Step Creation Handlers ---
        private static object CreateDelayStep(ScriptableObject graph, JObject @params, string prompt)
        {
            float delayTime = @params["delayTime"]?.ToObject<float?>() ?? 1.0f;
            string stepId = @params["stepId"]?.ToString();
            string tooltip = @params["tooltip"]?.ToString();
            Vector2 position = ParseVector2(@params["position"]) ?? Vector2.zero;

            // Parse delay time from prompt if not provided
            if (@params["delayTime"] == null && !string.IsNullOrEmpty(prompt))
            {
                var delayMatch = DelayPattern.Match(prompt);
                if (delayMatch.Success && float.TryParse(delayMatch.Groups[1].Value, out float parsedDelay))
                {
                    delayTime = parsedDelay;
                    Debug.Log($"Parsed delayTime from prompt: {delayTime}");
                }
            }

            if (delayTime < 0)
            {
                throw new ArgumentException("delayTime must be non-negative");
            }

            var delayStep = CreateStepNode(graph, "DelayStep", stepId);
            if (delayStep != null)
            {
                SetNodeProperty(delayStep, "position", position);
                SetNodeProperty(delayStep, "tooltip", tooltip ?? $"Wait for {delayTime} seconds");
                SetNodeProperty(delayStep, "timeToWait", delayTime, BindingFlags.NonPublic | BindingFlags.Instance);

                return new Dictionary<string, object>
                {
                    { "stepId", GetNodeProperty(delayStep, "stepId") ?? "unknown" },
                    { "stepType", "delay" },
                    { "delayTime", delayTime },
                    { "position", new { x = position.x, y = position.y } },
                    { "message", "Delay step created successfully" }
                };
            }
            return null;
        }

        private static object CreateClickStep(ScriptableObject graph, JObject @params, string prompt)
        {
            string targetName = @params["targetName"]?.ToString();
            string stepId = @params["stepId"]?.ToString();
            Vector2 position = ParseVector2(@params["position"]) ?? Vector2.zero;

            // Parse target from prompt if not provided
            if (string.IsNullOrEmpty(targetName) && !string.IsNullOrEmpty(prompt))
            {
                var clickMatch = ClickPattern.Match(prompt);
                if (clickMatch.Success)
                {
                    string prefabName = clickMatch.Groups[1].Value;
                    string childName = clickMatch.Groups[2].Success ? clickMatch.Groups[2].Value : prefabName;
                    targetName = $"{prefabName}/{childName}";
                    Debug.Log($"Parsed targetName from prompt: {targetName}");
                }
            }

            if (string.IsNullOrEmpty(targetName))
            {
                throw new ArgumentException("targetName parameter is required for click steps");
            }

            string[] targetParts = targetName.Split('/');
            if (targetParts.Length != 2)
            {
                throw new ArgumentException($"targetName must be in format 'prefabName/childName', got: {targetName}");
            }

            var clickStep = CreateStepNode(graph, "ClickStep", stepId);
            if (clickStep != null)
            {
                SetNodeProperty(clickStep, "position", position);

                // Create registry item for the target
                var registryItemType = FindType("RegistryItem");
                if (registryItemType != null)
                {
                    var registryItem = Activator.CreateInstance(registryItemType);
                    SetObjectProperty(registryItem, "prefabName", targetParts[0]);
                    SetObjectProperty(registryItem, "childName", targetParts[1]);
                    SetNodeProperty(clickStep, "_targetName", registryItem);
                }

                return new Dictionary<string, object>
                {
                    { "stepId", GetNodeProperty(clickStep, "stepId") ?? "unknown" },
                    { "stepType", "click" },
                    { "targetName", targetName },
                    { "position", new { x = position.x, y = position.y } },
                    { "message", "Click step created successfully" }
                };
            }
            return null;
        }

        // --- Helper Methods ---
        private static string SanitizeAssetPath(string path)
        {
            if (string.IsNullOrEmpty(path))
                return path;
            path = path.Replace('\\', '/');
            if (!path.StartsWith("Assets/", StringComparison.OrdinalIgnoreCase))
            {
                return "Assets/" + path.TrimStart('/');
            }
            return path;
        }

        private static bool IsStepsGraph(ScriptableObject graph)
        {
            InitializeTypes();
            return graph != null && stepsGraphType != null && stepsGraphType.IsInstanceOfType(graph);
        }

        private static void InitializeTypes()
        {
            if (stepsGraphType == null || nodeGraphType == null)
            {
                foreach (var assembly in AppDomain.CurrentDomain.GetAssemblies())
                {
                    try
                    {
                        var types = assembly.GetTypes();

                        if (nodeGraphType == null)
                            nodeGraphType = Array.Find(types, t => t.Name == "NodeGraph");

                        if (stepsGraphType == null)
                            stepsGraphType = Array.Find(types, t => t.Name == "StepsGraph");

                        if (nodeGraphType != null && stepsGraphType != null)
                            break;
                    }
                    catch
                    {
                        continue;
                    }
                }
            }
        }

        private static Type FindType(string typeName)
        {
            if (stepTypes.TryGetValue(typeName, out var cachedType))
                return cachedType;

            foreach (var assembly in AppDomain.CurrentDomain.GetAssemblies())
            {
                try
                {
                    var type = assembly.GetType(typeName, false, true) ??
                              Array.Find(assembly.GetTypes(), t => t.Name == typeName);
                    
                    if (type != null)
                    {
                        stepTypes[typeName] = type;
                        return type;
                    }
                }
                catch
                {
                    continue;
                }
            }
            return null;
        }

        private static ScriptableObject CreateStepNode(ScriptableObject graph, string stepTypeName, string stepId = null)
        {
            InitializeTypes();
            
            var stepType = FindType(stepTypeName);
            if (stepType == null)
            {
                Debug.LogError($"{stepTypeName} type not found!");
                return null;
            }

            if (nodeGraphType == null)
            {
                Debug.LogError("NodeGraph type not found!");
                return null;
            }

#if UNITY_EDITOR
            try
            {
                Undo.RecordObject(graph, $"Add {stepTypeName}");

                var addNodeMethod = nodeGraphType.GetMethod("AddNode", new Type[0]);
                if (addNodeMethod == null)
                {
                    Debug.LogError("AddNode method not found on NodeGraph!");
                    return null;
                }

                var genericAddNodeMethod = addNodeMethod.MakeGenericMethod(stepType);
                var stepNode = genericAddNodeMethod.Invoke(graph, null) as ScriptableObject;

                if (stepNode != null)
                {
                    // Set step ID
                    string finalStepId = string.IsNullOrEmpty(stepId) ? GenerateUniqueStepId(graph) : stepId;
                    SetNodeProperty(stepNode, "stepId", finalStepId);

                    // Register for undo and mark dirty
                    Undo.RegisterCreatedObjectUndo(stepNode, $"Add {stepTypeName}");
                    
                    string assetPath = AssetDatabase.GetAssetPath(graph);
                    if (!string.IsNullOrEmpty(assetPath))
                    {
                        AssetDatabase.AddObjectToAsset(stepNode, graph);
                        AssetDatabase.SaveAssets();
                    }
                    
                    EditorUtility.SetDirty(graph);
                    return stepNode;
                }
            }
            catch (Exception ex)
            {
                Debug.LogError($"Error creating {stepTypeName}: {ex.Message}");
            }
#endif
            return null;
        }

        private static bool SetNodeProperty(object target, string propertyName, object value, BindingFlags flags = BindingFlags.Public | BindingFlags.Instance)
        {
            try
            {
                var field = target.GetType().GetField(propertyName, flags);
                if (field != null)
                {
                    field.SetValue(target, value);
                    return true;
                }
                
                var property = target.GetType().GetProperty(propertyName, flags);
                if (property != null && property.CanWrite)
                {
                    property.SetValue(target, value);
                    return true;
                }
            }
            catch (Exception ex)
            {
                Debug.LogWarning($"Failed to set property '{propertyName}': {ex.Message}");
            }
            return false;
        }

        private static object GetNodeProperty(object target, string propertyName, BindingFlags flags = BindingFlags.Public | BindingFlags.Instance)
        {
            try
            {
                var field = target.GetType().GetField(propertyName, flags);
                if (field != null)
                {
                    return field.GetValue(target);
                }
                
                var property = target.GetType().GetProperty(propertyName, flags);
                if (property != null && property.CanRead)
                {
                    return property.GetValue(target);
                }
            }
            catch (Exception ex)
            {
                Debug.LogWarning($"Failed to get property '{propertyName}': {ex.Message}");
            }
            return null;
        }

        private static bool SetObjectProperty(object target, string propertyName, object value)
        {
            return SetNodeProperty(target, propertyName, value);
        }

        private static Vector2? ParseVector2(JToken token)
        {
            if (token == null)
                return null;

            try
            {
                if (token is JArray arr && arr.Count >= 2)
                {
                    return new Vector2(arr[0].ToObject<float>(), arr[1].ToObject<float>());
                }
                else if (token is JObject obj)
                {
                    float x = obj["x"]?.ToObject<float>() ?? 0f;
                    float y = obj["y"]?.ToObject<float>() ?? 0f;
                    return new Vector2(x, y);
                }
            }
            catch (Exception ex)
            {
                Debug.LogWarning($"Error parsing Vector2: {ex.Message}");
            }
            return null;
        }

        private static string GenerateUniqueStepId(ScriptableObject graph)
        {
            try
            {
                var nodesField = graph.GetType().GetField("nodes", BindingFlags.Public | BindingFlags.Instance);
                if (nodesField?.GetValue(graph) is System.Collections.IEnumerable nodes)
                {
                    int maxId = 0;
                    foreach (var node in nodes)
                    {
                        if (node is ScriptableObject so)
                        {
                            string id = GetNodeProperty(so, "stepId") as string;
                            if (!string.IsNullOrEmpty(id) && id.StartsWith("s") && int.TryParse(id.Substring(1), out int numericId))
                            {
                                maxId = Math.Max(maxId, numericId);
                            }
                        }
                    }
                    return "s" + (maxId + 1);
                }
            }
            catch (Exception ex)
            {
                Debug.LogWarning($"Error generating unique step ID: {ex.Message}");
            }
            return "s1";
        }

        // --- Helper Classes ---
        private class StepTypeHandler
        {
            public string TypeName { get; }
            public Func<ScriptableObject, JObject, string, object> CreateStep { get; }
            public string[] Keywords { get; }

            public StepTypeHandler(string typeName, Func<ScriptableObject, JObject, string, object> createStep, string[] keywords)
            {
                TypeName = typeName;
                CreateStep = createStep;
                Keywords = keywords;
            }
        }
    }
}

File: UnityMcpServer/src/tools/add_step.py
from typing import Any, Dict, Optional, List
from mcp.server.fastmcp import Context, FastMCP
from unity_connection import get_unity_connection
import re

def register_add_step_tools(mcp: FastMCP):
    """Register generic add step tools with the MCP server."""
    
    @mcp.tool()
    def add_step(
        ctx: Context,
        prompt: str = None,
        step_type: str = None,
        target_graph_path: str = None,
        delay_time: float = None,
        target_name: str = None,
        step_id: str = None,
        position: List[float] = None,
        tooltip: str = None
    ) -> Dict[str, Any]:
        """Adds a step to a StepsGraph with natural language support.
       
        Args:
            prompt: Natural language description of the step to add (optional)
            step_type: Type of step ('delay', 'click', etc.)
            target_graph_path: Path to the StepsGraph asset
            delay_time: Delay time in seconds (for delay steps)
            target_name: Target object name in format 'prefabName/childName' (for click steps)
            step_id: Optional step ID
            position: Optional position as [x, y]
            tooltip: Optional tooltip text
           
        Returns:
            Dictionary with results ('success', 'message', 'data')
        """
        return handle_add_step({
            "prompt": prompt,
            "step_type": step_type,
            "target_graph_path": target_graph_path,
            "delay_time": delay_time,
            "target_name": target_name,
            "step_id": step_id,
            "position": position,
            "tooltip": tooltip
        })

class StepParser:
    """Handles parsing of natural language prompts for step creation."""
    
    # Step type patterns and keywords
    STEP_TYPES = {
        'delay': {
            'keywords': ['delay', 'wait', 'pause', 'sleep'],
            'patterns': [
                r'(?:wait|delay|for:?\s*)?(\d+(?:\.\d+)?)\s*(?:seconds?|s)?',
                r'(\d+(?:\.\d+)?)\s*(?:second|sec|s)\s*(?:wait|delay)?',
                r'pause\s*(?:for)?\s*(\d+(?:\.\d+)?)',
            ]
        },
        'click': {
            'keywords': ['click', 'tap', 'press', 'button'],
            'patterns': [
                r'(?:click|tap|press)\s+(?:on\s+)?([\w-]+)(?:/([^/\s]+))?',
                r'(?:button|target)\s+([\w-]+)(?:/([^/\s]+))?',
            ]
        }
    }
    
    # Common patterns
    GRAPH_PATTERN = re.compile(r'(?:in|graph|asset:?\s*)?([\w-]+\.asset)', re.IGNORECASE)
    STEP_TYPE_PATTERN = re.compile(r'(?:add|create)\s+(?:a\s+)?(\w+)\s*(?:step)?', re.IGNORECASE)
    
    @classmethod
    def parse_prompt(cls, prompt: str) -> Dict[str, Any]:
        """Parse a natural language prompt to extract step parameters."""
        if not prompt:
            return {}
            
        result = {}
        prompt_lower = prompt.lower()
        
        # Extract graph path
        graph_match = cls.GRAPH_PATTERN.search(prompt)
        if graph_match:
            result['target_graph_path'] = graph_match.group(1)
        
        # Determine step type
        step_type = cls._determine_step_type(prompt_lower)
        if step_type:
            result['step_type'] = step_type
            
            # Parse type-specific parameters
            if step_type == 'delay':
                delay_time = cls._parse_delay_time(prompt)
                if delay_time is not None:
                    result['delay_time'] = delay_time
                    
            elif step_type == 'click':
                target_name = cls._parse_click_target(prompt)
                if target_name:
                    result['target_name'] = target_name
        
        return result
    
    @classmethod
    def _determine_step_type(cls, prompt_lower: str) -> Optional[str]:
        """Determine step type from prompt keywords."""
        # Check keywords for each step type
        for step_type, config in cls.STEP_TYPES.items():
            if any(keyword in prompt_lower for keyword in config['keywords']):
                return step_type
        
        # Fallback: try "add [type] step" pattern
        step_match = cls.STEP_TYPE_PATTERN.search(prompt_lower)
        if step_match:
            extracted_type = step_match.group(1).lower()
            if extracted_type in cls.STEP_TYPES:
                return extracted_type
                
        return None
    
    @classmethod
    def _parse_delay_time(cls, prompt: str) -> Optional[float]:
        """Parse delay time from prompt."""
        for pattern in cls.STEP_TYPES['delay']['patterns']:
            match = re.search(pattern, prompt, re.IGNORECASE)
            if match:
                try:
                    return float(match.group(1))
                except (ValueError, IndexError):
                    continue
        return None
    
    @classmethod
    def _parse_click_target(cls, prompt: str) -> Optional[str]:
        """Parse click target from prompt."""
        for pattern in cls.STEP_TYPES['click']['patterns']:
            match = re.search(pattern, prompt, re.IGNORECASE)
            if match:
                try:
                    prefab_name = match.group(1)
                    child_name = match.group(2) if match.lastindex > 1 and match.group(2) else prefab_name
                    return f"{prefab_name}/{child_name}"
                except IndexError:
                    continue
        return None

def handle_add_step(params: Dict[str, Any]) -> Dict[str, Any]:
    """
    Handle generic add step tool requests.
    
    Args:
        params: Dictionary containing step parameters
    
    Returns:
        Dictionary containing the response from Unity
    """
    try:
        prompt = params.get('prompt')
        step_type = params.get('step_type')
        target_graph_path = params.get('target_graph_path')
        
        # Parse prompt if provided and parameters are missing
        if prompt:
            parsed_params = StepParser.parse_prompt(prompt)
            
            # Use parsed values if original parameters are missing
            if not step_type and 'step_type' in parsed_params:
                step_type = parsed_params['step_type']
            if not target_graph_path and 'target_graph_path' in parsed_params:
                target_graph_path = parsed_params['target_graph_path']
            
            # Merge type-specific parsed parameters
            for key, value in parsed_params.items():
                if key not in ['step_type', 'target_graph_path'] and params.get(key) is None:
                    params[key] = value
        
        # Validate required parameters
        if not step_type:
            available_types = ', '.join(StepParser.STEP_TYPES.keys())
            return {
                "success": False,
                "message": f"Could not determine step type from prompt. Please specify 'step_type' parameter. Available types: {available_types}"
            }
        
        if not target_graph_path:
            return {
                "success": False,
                "message": "target_graph_path parameter is required"
            }
        
        # Validate step type
        if step_type not in StepParser.STEP_TYPES:
            available_types = ', '.join(StepParser.STEP_TYPES.keys())
            return {
                "success": False,
                "message": f"Unknown step type: '{step_type}'. Available types: {available_types}"
            }
        
        # Type-specific validation
        validation_result = _validate_step_params(step_type, params)
        if not validation_result['success']:
            return validation_result
        
        # Prepare Unity parameters
        unity_params = {
            "stepType": step_type,
            "targetGraphPath": target_graph_path
        }
        
        # Add optional parameters
        for param_name, unity_name in [
            ('delay_time', 'delayTime'),
            ('target_name', 'targetName'),
            ('step_id', 'stepId'),
            ('position', 'position'),
            ('tooltip', 'tooltip'),
            ('prompt', 'prompt')  # Include original prompt for Unity-side parsing
        ]:
            if params.get(param_name) is not None:
                unity_params[unity_name] = params[param_name]
        
        # Send command to Unity (assuming generic handler exists)
        response = get_unity_connection().send_command("add_step", unity_params)
        
        # Process response from Unity
        if response.get("success"):
            return {
                "success": True,
                "message": f"{step_type.capitalize()} step added successfully",
                "data": response.get("data")
            }
        else:
            return {
                "success": False,
                "message": response.get("error", f"Failed to add {step_type} step")
            }
            
    except Exception as e:
        return {
            "success": False,
            "message": f"Error adding step: {str(e)}"
        }

def _validate_step_params(step_type: str, params: Dict[str, Any]) -> Dict[str, Any]:
    """Validate step-specific parameters."""
    
    if step_type == 'delay':
        delay_time = params.get('delay_time')
        if delay_time is None:
            return {
                "success": False,
                "message": "delay_time parameter is required for delay steps"
            }
        if delay_time < 0:
            return {
                "success": False,
                "message": "delay_time must be non-negative"
            }
            
    elif step_type == 'click':
        target_name = params.get('target_name')
        if not target_name:
            return {
                "success": False,
                "message": "target_name parameter is required for click steps"
            }
        if '/' not in target_name:
            return {
                "success": False,
                "message": "target_name must be in format 'prefabName/childName'"
            }
    
    return {"success": True}

# Backwards compatibility - keep individual tool functions
def register_add_delay_step_tools(mcp: FastMCP):
    """Register delay step tools for backwards compatibility."""
    
    @mcp.tool()
    def add_delay_step(
        ctx: Context,
        prompt: str = None,
        target_graph_path: str = None,
        delay_time: float = None,
        step_id: str = None
    ) -> Dict[str, Any]:
        """Adds a delay step to a StepsGraph with natural language support."""
        return handle_add_step({
            "prompt": prompt,
            "step_type": "delay",
            "target_graph_path": target_graph_path,
            "delay_time": delay_time,
            "step_id": step_id
        })

def register_add_click_step_tools(mcp: FastMCP):
    """Register click step tools for backwards compatibility."""
    
    @mcp.tool()
    def add_click_step(
        ctx: Context,
        prompt: str = None,
        target_graph_path: str = None,
        target_name: str = None,
        step_id: str = None
    ) -> Dict[str, Any]:
        """Adds a click step to a StepsGraph with natural language support."""
        return handle_add_step({
            "prompt": prompt,
            "step_type": "click",
            "target_graph_path": target_graph_path,
            "target_name": target_name,
            "step_id": step_id
        })